<?xml version="1.0"?>
<sdf version="1.4">
	<world name="default">

		<!-- Ground Plane -->
		<include>
			<uri>model://ground_plane</uri>
		</include>

		<include>
			<uri>model://sun</uri>
		</include>

		<model name="ProjectBot">
			<pose>0 0 0 0 0 0</pose>
			<link name='chassis'>
				<inertial>
					<mass>13.741</mass>
					<pose>-.269696 -.006844 .12974 0 0 0</pose>
				</inertial>
				<collision name='collision1'>
					<pose>-.1285 0 .12 0 0 0</pose>
					<geometry>
						<box>
							<size>.217 .476 .205</size>
						</box>
					</geometry>
				</collision>
				<visual name='visual1'>
					<pose>-.1285 0 .12 0 0 0</pose>
					<geometry>
						<box>
							<size>.217 .476 .205</size>
						</box>
					</geometry>
				</visual>
				<collision name='collision2'>
					<pose>-.01 0 .12 0 0 0</pose>
					<geometry>
						<box>
							<size>.02 .6 .205</size>
						</box>
					</geometry>
				</collision>
				<visual name='visual2'>
					<pose>-.01 0 .12 0 0 0</pose>
					<geometry>
						<box>
							<size>.02 .6 .205</size>
						</box>
					</geometry>
				</visual>
				<collision name='collision3'>
					<pose>-.4185 0 .12 0 0 0</pose>
					<geometry>
						<box>
							<size>.363 .6 .205</size>
						</box>
					</geometry>
				</collision>
				<visual name='visual3'>
					<pose>-.4185 0 .12 0 0 0</pose>
					<geometry>
						<box>
							<size>.363 .6 .205</size>
						</box>
					</geometry>
				</visual>
				<collision name='collision4'>
					<pose>-.3 0 .2425 0 0 0</pose>
					<geometry>
						<box>
							<size>.6 .6 .04</size>
						</box>
					</geometry>
				</collision>
				<visual name='visual4'>
					<pose>-.3 0 .2425 0 0 0</pose>
					<geometry>
						<box>
							<size>.6 .6 .04</size>
						</box>
					</geometry>
				</visual>
				<collision name='caster_collision1'>
				    	<pose>-0.5635 .2635 .0175 0 0 0</pose>
				        <geometry>
						<sphere>
							<radius>.0175</radius>
				      		</sphere>
				    	</geometry>

				        <surface>
				      		<friction>
							<ode>
							  <mu>0</mu>
							  <mu2>0</mu2>
							  <slip1>1.0</slip1>
							  <slip2>1.0</slip2>
							</ode>
				      		</friction>
				    	</surface>
			   	</collision>
			  	<visual name='caster_visual1'>
					<pose>-0.5635 .2635 .0175 0 0 0</pose>
				        <geometry>
						<sphere>
							<radius>.0175</radius>
				      		</sphere>
				    	</geometry>
			  	</visual>
				<collision name='caster_collision2'>
				    	<pose>-0.5635 -.2635 .0175 0 0 0</pose>
				        <geometry>
						<sphere>
							<radius>.0175</radius>
				      		</sphere>
				    	</geometry>

				        <surface>
				      		<friction>
							<ode>
							  <mu>0</mu>
							  <mu2>0</mu2>
							  <slip1>1.0</slip1>
							  <slip2>1.0</slip2>
							</ode>
				      		</friction>
				    	</surface>
			   	</collision>
			  	<visual name='caster_visual2'>
					<pose>-0.5635 -.2635 .0175 0 0 0</pose>
				        <geometry>
						<sphere>
							<radius>.0175</radius>
				      		</sphere>
				    	</geometry>
			  	</visual>
			</link>
			<link name="wheelLeft">
				<pose>-.125 .263 .1 1.5707 0 0</pose>
				<inertial>
					<mass>1.789</mass>
					<pose>0 0 .002728 0 0 0</pose>
				</inertial>
				<collision name="collisionLW">
				        <geometry>
						<cylinder>
							<radius>.1</radius>
							<length>.05</length>
				      		</cylinder>
				    	</geometry>
				</collision>
				<visual name="visualLW">
				        <geometry>
						<cylinder>
							<radius>.1</radius>
							<length>.05</length>
				      		</cylinder>
				    	</geometry>
				</visual>
			</link>
			<link name="wheelRight">
				<pose>-.125 -.263 .1 1.5707 0 0</pose>
				<inertial>
					<mass>1.789</mass>
					<pose>0 0 -.002728 0 0 0</pose>
				</inertial>
				<collision name="collisionRW">
				        <geometry>
						<cylinder>
							<radius>.1</radius>
							<length>.05</length>
				      		</cylinder>
				    	</geometry>
				</collision>
				<visual name="visualRW">
				        <geometry>
						<cylinder>
							<radius>.1</radius>
							<length>.05</length>
				      		</cylinder>
				    	</geometry>
				</visual>
			</link>
			<joint type="revolute" name="left_wheel_joint">

				<child>wheelLeft</child>
				<parent>chassis</parent>
				<axis>
					<xyz>0 -1 0</xyz>
				</axis>
			</joint>
			<joint type="revolute" name="right_wheel_joint">

				<child>wheelRight</child>
				<parent>chassis</parent>
				<axis>
					<xyz>0 -1 0</xyz>
				</axis>
			</joint>
			<plugin name="differential_drive_controller" filename="libgazebo_ros_diff_drive.so">
				<alwaysOn>true</alwaysOn>
				<updateRate>20</updateRate>
				<leftJoint>left_wheel_joint</leftJoint>
				<rightJoint>right_wheel_joint</rightJoint>
				<legacyMode>true</legacyMode>
				<wheelSeparation>0.526</wheelSeparation>
				<wheelDiameter>0.2</wheelDiameter>
				<torque>20</torque>
				<robotNamespace>/</robotNamespace>
				<commandTopic>cmd_vel</commandTopic>
				<odometryTopic>odom</odometryTopic>
				<odometryFrame>odom</odometryFrame>
				<robotBaseFrame>base_footprint</robotBaseFrame>
			</plugin>								
			<plugin name="EncoderPlugin" filename="libencoder.so"/>
		</model>
		<physics type='ode'>
			<real_time_factor>1</real_time_factor>
			<max_step_size>0.001</max_step_size>
  			<real_time_update_rate>1000.000000</real_time_update_rate>
			<gravity>0.000000 0.000000 -9.800000</gravity>
		</physics>
	</world>
</sdf>
